cpp/src/arrow/csv/writer.cc

```diff
@@ -222,13 +257,15 @@ class UnquotedColumnPopulator : public ColumnPopulator {
       return Status::OK();
     };
 
-    return VisitArraySpanInline<StringType>(*casted_array_->data(), valid_function,
-                                            null_function);
+    auto casted_array = checked_pointer_cast<StringArrayType>(array_);
+    return VisitArraySpanInline<typename StringArrayType::TypeClass>(
+        *casted_array->data(), valid_function, null_function);
```

Comment:
I don't think the pointer cast is needed here?
```suggestion
    return VisitArraySpanInline<typename StringArrayType::TypeClass>(
        array_->data(), valid_function, null_function);
```

Reply:
Applied. Thanks!